### PR TITLE
handle json errors

### PR DESCRIPTION
**Pull request by achille-roussel**
@segmentio/infra 

This should help us get more insights when unsupported values are set to in the log data.

### CODE REVIEW

File: logger.go [hunk 1]
```diff
@@ -21,5 +22,22 @@ func NewLogger(w io.Writer) Logger {
...
+	if err = enc.Encode(event); err == nil {
+		return
+	}
+
+	// Attempts to recover from invalid data put in the free form Event.Data field.
+	switch err.(type) {
+	case *json.UnsupportedTypeError, *json.UnsupportedValueError, *json.MarshalerError:
+		event.Level = ALERT
+		event.Info.Errors = append(event.Info.Errors, MakeEventError(err))
+		event.Data = EventData{"unserializable": fmt.Sprintf("%#v", event.Data)}
```
Comment (yields) [inline logger.go]: are there any size checks? this string could be huge, we might wanna truncate it?

**Comment by yields** [comment]
one comment, otherwise shipit! :)

File: logger.go (hunk 1 above)
Comment (achille-roussel) [inline logger.go]: The only limit I'm aware of is 1 MB per event batch when sending to CloudWatchLogs... I don't think we'll ever reach this with a single event tho.

**Comment by calvinfo** [comment]
Nice!

**Pull request merged**